### PR TITLE
[WIP] Add org/repo level pod timeout

### PR DESCRIPTION
**Pull request by ArkaSaha30**
What type of PR is this?
/kind feature

What this PR does / why we need it:
Allow each team control their PodPendingTimeout, PodRunningTimeout, and PodUnscheduledTimeout at org/repo level.
These values are configured globally, and might not be most optimal for all teams, adjusting the values of these fields often need to get approvals from lots of stakeholders when the prow instance has many tenants.

Which issue(s) this PR fixes:
Refers to: https://github.com/kubernetes/test-infra/issues/26614

### CODE REVIEW

File: prow/config/config.go [hunk 1]
```diff
@@ -698,14 +708,23 @@ func matches(givenOrgRepo, givenCluster, orgRepo, cluster string) bool {
 	return orgRepoMatch && clusterMatch
 }
 
+func matchesDefaultDecorationConfigEntry(givenOrgRepo, givenCluster, orgRepo, cluster string, givenPodPendingTimeout, givenPodRunningTimeout, givenPodUnscheduledTimeout, podPendingTimeout, podRunningTimeout, podUnscheduledTimeout *metav1.Duration) bool {
```
Comment (stevekuznetsov) [inline prow/config/config.go:711]: Why do the timeouts have to match if the org/repo does?

File: prow/config/config.go [hunk 2]
```diff
@@ -704,7 +714,7 @@ func (d *ProwJobDefaultEntry) matches(repo, cluster string) bool {
 }
 
 // matches returns true iff all the filters for the entry match a job.
-func (d *DefaultDecorationConfigEntry) matches(repo, cluster string) bool {
+func (d *DefaultDecorationConfigEntry) matches(repo, cluster string, podPendingTimeout, podRunningTimeout, podUnscheduledTimeout *metav1.Duration) bool {
```
Comment (stevekuznetsov) [inline prow/config/config.go:717]: This diff is not needed.

File: prow/config/config.go [hunk 3]
```diff
@@ -736,10 +746,10 @@ func (pc *ProwConfig) mergeProwJobDefault(repo, cluster string, jobDefault *prow
 // mergeDefaultDecorationConfig finds all matching DefaultDecorationConfigEntry
 // for a job and merges them sequentially before merging into the job's own
 // DecorationConfig. Configs merged later override values from earlier configs.
-func (p *Plank) mergeDefaultDecorationConfig(repo, cluster string, jobDC *prowapi.DecorationConfig) *prowapi.DecorationConfig {
+func (p *Plank) mergeDefaultDecorationConfig(repo, cluster string, podPendingTimeout, podRunningTimeout, podUnscheduledTimeout *metav1.Duration, jobDC *prowapi.DecorationConfig) *prowapi.DecorationConfig {
 	var merged *prowapi.DecorationConfig
 	for _, entry := range p.DefaultDecorationConfigs {
-		if entry.matches(repo, cluster) {
+		if entry.matches(repo, cluster, podPendingTimeout, podRunningTimeout, podUnscheduledTimeout) {
```
Comment (stevekuznetsov) [inline prow/config/config.go:752]: This diff is not needed.

File: prow/config/config.go [hunk 4]
```diff
@@ -736,10 +746,10 @@ func (pc *ProwConfig) mergeProwJobDefault(repo, cluster string, jobDefault *prow
 // mergeDefaultDecorationConfig finds all matching DefaultDecorationConfigEntry
 // for a job and merges them sequentially before merging into the job's own
 // DecorationConfig. Configs merged later override values from earlier configs.
-func (p *Plank) mergeDefaultDecorationConfig(repo, cluster string, jobDC *prowapi.DecorationConfig) *prowapi.DecorationConfig {
+func (p *Plank) mergeDefaultDecorationConfig(repo, cluster string, podPendingTimeout, podRunningTimeout, podUnscheduledTimeout *metav1.Duration, jobDC *prowapi.DecorationConfig) *prowapi.DecorationConfig {
```
Comment (stevekuznetsov) [inline prow/config/config.go:749]: This diff is not needed.

File: prow/config/config.go [hunk 5]
```diff
@@ -760,7 +770,7 @@ func (c *Config) GetProwJobDefault(repo, cluster string) *prowapi.ProwJobDefault
 // config for a given repo and cluster. It is primarily used for best effort
 // guesses about GCS configuration for undecorated jobs.
 func (p *Plank) GuessDefaultDecorationConfig(repo, cluster string) *prowapi.DecorationConfig {
-	return p.mergeDefaultDecorationConfig(repo, cluster, nil)
+	return p.mergeDefaultDecorationConfig(repo, cluster, nil, nil, nil, nil)
```
Comment (stevekuznetsov) [inline prow/config/config.go:773]: This diff is not needed.

File: prow/config/config.go [hunk 6]
```diff
@@ -770,9 +780,12 @@ func defaultDecorationMapToSlice(m map[string]*prowapi.DecorationConfig) []*Defa
 	var entries []*DefaultDecorationConfigEntry
 	add := func(repo string, dc *prowapi.DecorationConfig) {
 		entries = append(entries, &DefaultDecorationConfigEntry{
-			OrgRepo: repo,
-			Cluster: "",
-			Config:  dc,
+			OrgRepo:               repo,
+			Cluster:               "",
+			Config:                dc,
+			PodPendingTimeout:     nil,
```
Comment (stevekuznetsov) [inline prow/config/config.go:786]: `nil` is the default value, you do not need to specify it

File: prow/config/config.go [hunk 7]
```diff
@@ -1890,15 +1903,16 @@ func setPeriodicProwJobDefaults(c *Config, ps *Periodic) {
 
 	ps.ProwJobDefault = c.mergeProwJobDefault(repo, ps.Cluster, ps.ProwJobDefault)
 }
+
 func setPresubmitDecorationDefaults(c *Config, ps *Presubmit, repo string) {
 	if shouldDecorate(&c.JobConfig, &ps.JobBase.UtilityConfig) {
-		ps.DecorationConfig = c.Plank.mergeDefaultDecorationConfig(repo, ps.Cluster, ps.DecorationConfig)
+		ps.DecorationConfig = c.Plank.mergeDefaultDecorationConfig(repo, ps.Cluster, &metav1.Duration{Duration: 10 * time.Minute}, &metav1.Duration{Duration: 48 * time.Hour}, &metav1.Duration{Duration: 5 * time.Minute}, ps.DecorationConfig)
```
Comment (stevekuznetsov) [inline prow/config/config.go:1909]: Shouldn't the user be in charge the overall defaults?

File: prow/config/config.go [hunk 8]
```diff
@@ -1890,15 +1903,16 @@ func setPeriodicProwJobDefaults(c *Config, ps *Periodic) {
...
 	}
 }
 
 func setPostsubmitDecorationDefaults(c *Config, ps *Postsubmit, repo string) {
 	if shouldDecorate(&c.JobConfig, &ps.JobBase.UtilityConfig) {
-		ps.DecorationConfig = c.Plank.mergeDefaultDecorationConfig(repo, ps.Cluster, ps.DecorationConfig)
+		ps.DecorationConfig = c.Plank.mergeDefaultDecorationConfig(repo, ps.Cluster, nil, nil, nil, ps.DecorationConfig)
```
Comment (stevekuznetsov) [inline prow/config/config.go:1915]: Why are these `nil`?

File: prow/config/config.go [hunk 9]
```diff
@@ -1909,7 +1923,7 @@ func setPeriodicDecorationDefaults(c *Config, ps *Periodic) {
 			repo = fmt.Sprintf("%s/%s", ps.UtilityConfig.ExtraRefs[0].Org, ps.UtilityConfig.ExtraRefs[0].Repo)
 		}
 
-		ps.DecorationConfig = c.Plank.mergeDefaultDecorationConfig(repo, ps.Cluster, ps.DecorationConfig)
+		ps.DecorationConfig = c.Plank.mergeDefaultDecorationConfig(repo, ps.Cluster, nil, nil, nil, ps.DecorationConfig)
```
Comment (stevekuznetsov) [inline prow/config/config.go:1926]: Why are these nil?

File: prow/config/config.go [hunk 10]
```diff
@@ -2179,7 +2193,6 @@ func (c Config) validatePostsubmits(postsubmits []Postsubmit) error {
 
 // validatePeriodics validates a set of periodics.
 func (c Config) validatePeriodics(periodics []Periodic) error {
-
```
Comment (stevekuznetsov) [inline prow/config/config.go:2182]: Let's not introduce diffs like these that are not related to the code that's being changed.

File: prow/config/config.go [hunk 11]
```diff
@@ -2199,7 +2212,6 @@ func (c Config) validatePeriodics(periodics []Periodic) error {
 // ValidateJobConfig validates if all the jobspecs/presets are valid
 // if you are mutating the jobs, please add it to finalizeJobConfig above.
 func (c *Config) ValidateJobConfig() error {
-
```
Comment (stevekuznetsov) [inline prow/config/config.go:2202]: Let's not introduce diffs like these that are not related to the code that's being changed.

File: prow/config/config.go [hunk 12]
```diff
@@ -2474,7 +2486,6 @@ func parseProwConfig(c *Config) error {
 	for name, templates := range c.Tide.MergeTemplate {
 		if templates.TitleTemplate != "" {
 			titleTemplate, err := template.New("CommitTitle").Parse(templates.TitleTemplate)
-
```
Comment (stevekuznetsov) [inline prow/config/config.go:2477]: Let's not introduce diffs like these that are not related to the code that's being changed.

File: prow/config/config.go [hunk 13]
```diff
@@ -2484,7 +2495,6 @@ func parseProwConfig(c *Config) error {
 
 		if templates.BodyTemplate != "" {
 			bodyTemplate, err := template.New("CommitBody").Parse(templates.BodyTemplate)
-
```
Comment (stevekuznetsov) [inline prow/config/config.go:2487]: Let's not introduce diffs like these that are not related to the code that's being changed.

**Review by stevekuznetsov — COMMENTED** [review]
Could we add some test cases for this new feature?

File: prow/config/config.go (hunk 2 above)
Comment (ArkaSaha30) [reply prow/config/config.go:717]: Removed 👍

File: prow/config/config.go (hunk 4 above)
Comment (ArkaSaha30) [reply prow/config/config.go:749]: Removed 👍

File: prow/config/config.go (hunk 3 above)
Comment (ArkaSaha30) [reply prow/config/config.go:752]: Removed 👍

File: prow/config/config.go (hunk 5 above)
Comment (ArkaSaha30) [reply prow/config/config.go:773]: Removed 👍

File: prow/config/config.go (hunk 6 above)
Comment (ArkaSaha30) [reply prow/config/config.go:786]: Removed 👍

File: prow/config/config.go (hunk 7 above)
Comment (ArkaSaha30) [reply prow/config/config.go:1909]: Removed 👍

File: prow/config/config.go (hunk 8 above)
Comment (ArkaSaha30) [reply prow/config/config.go:1915]: Removed 👍

File: prow/config/config.go (hunk 9 above)
Comment (ArkaSaha30) [reply prow/config/config.go:1926]: Removed 👍

File: prow/config/config.go (hunk 10 above)
Comment (ArkaSaha30) [reply prow/config/config.go:2182]: Removed 👍

File: prow/config/config.go (hunk 11 above)
Comment (ArkaSaha30) [reply prow/config/config.go:2202]: Removed 👍

File: prow/config/config.go (hunk 12 above)
Comment (ArkaSaha30) [reply prow/config/config.go:2477]: Removed 👍

File: prow/config/config.go (hunk 13 above)
Comment (ArkaSaha30) [reply prow/config/config.go:2487]: Removed 👍

File: prow/config/config.go [hunk 14]
```diff
@@ -688,6 +688,16 @@ type DefaultDecorationConfigEntry struct {
 	// by sequentially merging with later entries overriding fields from earlier
 	// entries.
 	Config *prowapi.DecorationConfig `json:"config,omitempty"`
+
+	// PodPendingTimeout is after how long the controller will perform a garbage
```
Comment (chaodaiG) [inline prow/config/config.go:692]: These are unused, please remove

**Review by chaodaiG — COMMENTED** [review]
please add unit test to ensure that both existing configuration and the new configuration would work. In addition, logics are needed for merging the configuration when the timeouts are defined on all 3 levels: global, per repo, on job, which I'm not seeing here, is this already magically taken care of?

File: prow/config/config.go (hunk 14 above)
Comment (ArkaSaha30) [reply prow/config/config.go]: Sure, let me check that 👍
